rows_test.go

```diff
@@ -0,0 +1,586 @@
+			want: []testAtomicTypesRowSmallInt{
+				{key: "max int8", testString: "hello", testBytes: []byte("hello"), testInt: 127, testFloat: 42, testBool: true},
+			},
+		},
+		{
+			name:      "read max uint8 into int8",
+			input:     `SELECT * FROM TestOverflowRead WHERE key = "max uint8"`,
+			wantError: wantError{scan: true},
+			want: []testAtomicTypesRowSmallInt{
+				{key: "max uint8", testString: "hello", testBytes: []byte("hello"), testInt: 0, testFloat: 0, testBool: false},
+			},
+		},
+	}
+
+	// Run tests.
+	for _, tc := range tests {
+
+		rows, err := db.QueryContext(ctx, tc.input)
+		if (err != nil) && (!tc.wantError.query) {
+			t.Errorf("%s: unexpected query error: %v", tc.name, err)
+		}
+		if (err == nil) && (tc.wantError.query) {
+			t.Errorf("%s: expected query error but error was %v", tc.name, err)
+		}
+
+		got := []testAtomicTypesRowSmallInt{}
+		for rows.Next() {
+			var curr testAtomicTypesRowSmallInt
+			err := rows.Scan(
+				&curr.key, &curr.testString, &curr.testBytes, &curr.testInt, &curr.testFloat, &curr.testBool)
+			if (err != nil) && (!tc.wantError.scan) {
+				t.Errorf("%s: unexpected scan error: %v", tc.name, err)
+			}
+			if (err == nil) && (tc.wantError.scan) {
+				t.Errorf("%s: expected scan error but error was %v", tc.name, err)
+			}
+
+			got = append(got, curr)
+		}
+
+		rows.Close()
+		err = rows.Err()
+		if (err != nil) && (!tc.wantError.close) {
+			t.Errorf("%s: unexpected rows.Err error: %v", tc.name, err)
+		}
+		if (err == nil) && (tc.wantError.close) {
+			t.Errorf("%s: expected rows.Err error but error was %v", tc.name, err)
+		}
+
+		if !reflect.DeepEqual(tc.want, got) {
+			t.Errorf("%s: unexpected rows. want: %v, got: %v", tc.name, tc.want, got)
+		}
+
+	}
+
+	// Drop table.
+	_, err = db.ExecContext(ctx, `DROP TABLE TestOverflowRead`)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+}
+
+func TestRowsAtomicTypePermute(t *testing.T) {
+
+	// Open db.
+	ctx := context.Background()
+	db, err := sql.Open("spanner", dsn)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+
+	type then struct {
+		wantStr      string
+		wantErrorStr bool
+		wantByt      []byte
+		wantErrorByt bool
+		wantInt      int
+		wantErrorInt bool
+		wantFlo      float64
+		wantErrorFlo bool
+		wantBoo      bool
+		wantErrorBoo bool
+	}
+
+	tests := []struct {
+		name     string
+		typeName string
+		given    []string
+		when     string
+		then     then
+		tearDown string
+	}{
+		{
+			name:     "read spanner bytes (\"hello\") into go types",
+			typeName: "bytes",
+			given: []string{
+				`CREATE TABLE TestBadTypeInt (val BYTES(1024)) PRIMARY KEY (val)`,
+				`INSERT INTO TestBadTypeInt (val) VALUES (CAST("hello" as bytes)) `,
+			},
+			when: `SELECT * FROM TestBadTypeInt`,
+			then: then{
+				wantStr:      "hello",
+				wantByt:      []byte("hello"),
+				wantErrorInt: true,
+				wantErrorFlo: true,
+				wantErrorBoo: true,
+			},
+			tearDown: `DROP TABLE TestBadTypeInt`,
```

Comment:
This table name doesn't match the test case. The other test cases all use this table name. Please update these to either match the test case or to a more general table name.